acceptancetests/assess_lxd_space.py

```diff
@@ -0,0 +1,179 @@
+          in json format, in comparison to no space exists.
+       2. Total number of space should be 1.
+       3. The name of new added space should be the same as specified.
+       4. Total number of subnet CIDR from this new added space should be 1.
+       5. The CIDR data from new added space should be the same as specified.
+    """
+    space_name, subnet_cidr = add_space_with_existing_subnet(
+        client, space_name='testspace')
+    # New added space should exist at this point,
+    # set merge_stderr to False to eliminate noise in output.
+    raw_output = client.get_juju_output(
+        'spaces', '--format', 'json', include_e=False, merge_stderr=False)
+    try:
+        space_output = json.loads(raw_output)
+    except ValueError as e:
+        log.error('New added space cannot be found.')
+        raise e
+    # These 4 validations are in sequence.
+    if len(space_output['spaces'].keys()) != 1:
+        raise JujuAssertionError(
+            'Incorrect number of space(s). Found: {}; Expected: 1.'.format(
+            str(len(space_output['spaces'].keys()))))
+    elif space_output['spaces'].keys()[0] != space_name:
+        raise JujuAssertionError(
+            'Incorrect space name. Found: {}; Expected: {}.'.format(
+            space_output['spaces'].keys()[0], space_name))
+    elif len(space_output['spaces'][space_name].keys()) != 1:
+        raise JujuAssertionError(
+            'Incorrect number of CIDR(s). Found: {}; Expected: 1.'.format(
+            str(len(space_output['spaces'][space_name].keys()))))
+    elif space_output['spaces'][space_name].keys()[0] != subnet_cidr:
+        raise JujuAssertionError(
+            'Incorrect CIDR name. Found: {}; Expected: {}.'.format(
+            space_output['spaces'][space_name].keys()[0], subnet_cidr))
+    else:
+        log.info('New added space {} has been validated.'.format(space_name))
+
+
+def assert_app_status(client, charm_name, expected):
+    # Run juju status --format json
+    log.info('Checking current status of app {}.'.format(charm_name))
+    status_output = json.loads(
+        client.get_juju_output('status', '--format', 'json', include_e=False))
+    app_status = status_output['applications'][charm_name]['application-status']['current']
+
+    if app_status != expected:
+        raise JujuAssertionError(
+            'App status is incorrect. '\
+            'Found: {}\nExpected: {}\n.'.format(
+            app_status, expected))
+    else:
+        log.info('The current status of app {} is: {}; Expected: {}'.format(
+        charm_name, app_status, expected))
+
+
+def parse_args(argv):
+    """Parse all arguments."""
+    parser = argparse.ArgumentParser(
+        description="Test for lxd spaces and subnets constraints.")
+    add_basic_testing_arguments(parser)
+    return parser.parse_args(argv)
+
+
+def assess_lxd_space(client):
+    environment = os.getenv('ENV', default='parallel-aws')
```

Comment:
Not quite, the intent is fine the implementation is fine.

1. Setting a default means that anytime it's not set, regardless of what cloud it's bootstrapped on, the test won't fail early but try and continue the test. (i.e. ./assess_lxd_space.py parallel-azure /snap/bin/juju ... will bootstrap in azure).
2. If I set ENV to default-aws and try and run this test it will fail early even though it's bootstrapping in AWS.

The fix is to check which cloud is being used, not which credentials are being used. You can use client.env.get_cloud() before or after a bootstrap to get what cloud a client is setup to bootstrap in (i.e. check that cloud is 'aws').
This gives you the cloud, not credentials used.
 
There may be some confusion here.
Due to the peculiar way the jujupy test framework handles it's "environments" there is this idea of a "credential name" stored in environments.yaml (which isn't a juju 2 thing, but a juju 1 thing) the ModelClient/BootstrapManager constructor does a translation taking the env passed in, looking up in environments.yaml and setting everything up from there. So all 'parallel-aws' is is a look up key to which credentials to use for the bootstrap.